Update EIP-8141: approval effects are not reverted by frame revert or atomic-batch unroll (#11955)
ethereum/EIPs: EIPS/eip-8141.md

- `APPROVE_EXECUTION_AND_PAYMENT` performs both.

These effects take hold only if the frame that executed `APPROVE` completes successfully; if that frame reverts, they are discarded with it. Once committed, they are transaction-scoped: a subsequent frame revert or an atomic-batch unroll MUST NOT roll them back, and they persist for the remainder of the transaction, even if the `APPROVE` frame is a member of an atomic batch that is later unrolled. They remain part of the enclosing transaction and block state, so they are discarded only if the transaction is invalid or its block is removed by a reorg.

The nonce increment and cost collection are ordinary state writes. A client that implements atomic-batch rollback by restoring a snapshot taken at the start of the batch MUST exempt these writes, and the `payer` assignment, from that rollback (or re-apply them afterwards). Otherwise an in-batch `APPROVE_PAYMENT` whose batch is unrolled would have its debit and nonce reverted while `payer` survives, so the terminal `payer != None` check passes and the end-of-transaction refund pays the payer funds that were never collected.

> 

Reviewer: The approval effects are already listed in the behavior section for `APPROVE` so I don't think they are needed here.

The middle / last section is kind of wrong. VERIFY frames can't fail, otherwise the whole transaction is invalid. This part is probably a bit under specified since it only says "if the VERIFY frame reverts, the tx is invalid". Might just be simpler to disallow VERIFY frames in batches tbh.

4. `pay` must execute in `VERIFY` mode, have flags set to `APPROVE_PAYMENT`, and must successfully call `APPROVE(APPROVE_PAYMENT)`
5. No frame in the validation prefix may have the `ATOMIC_BATCH_FLAG` set.
6. The sum of `gas_limit` values across the validation prefix, plus the intrinsic cost of validating `tx.signatures`, must not exceed `MAX_VERIFY_GAS`.

> 7. Nodes should stop simulation immediately once `payer` has been set. As with the approval effects in the Behavior section, `payer` is set only when the frame that executes the payment-scoped `APPROVE` completes successfully; an `APPROVE` in an inner call does not set `payer` until that enclosing frame returns successfully, so simulation must continue through it.

Reviewer: ```suggestion
7. Nodes should stop simulation once `payer` has been set and the associated VERIFY frame completes successfully.
```
